internal/api/authorization/k8s.go

```diff
@@ -0,0 +1,639 @@
+
+	kubeClient, err := kubernetes.NewForConfig(kubeconfig)
+	if err != nil {
+		return nil, nil
+	}
+
+	sarClient := kubeClient.AuthorizationV1()
+	authorizer, err := newAuthorizer(sarClient)
+	if err != nil {
+		return nil, nil
+	}
+
+	tokenClient := kubeClient.AuthenticationV1()
+	kubernetesAuthenticator, err := newAuthenticator(
+		tokenClient,
+	)
+	if err != nil {
+		return nil, nil
+	}
+
+	return &k8sImpl{
+		Authenticator:    kubernetesAuthenticator,
+		Authorizer:       authorizer,
+		Kubeconfig:       kubeconfig,
+		kubeClient:       kubeClient,
+		guestPermissions: conf.Security.Authorization.GuestPermissions,
+	}, nil
+}
+
+type k8sImpl struct {
+	Authenticator    authenticator.Request
+	Authorizer       authorizer.Authorizer
+	Kubeconfig       *rest.Config
+	kubeClient       *kubernetes.Clientset
+	guestPermissions []*v1Role.Permission
+}
+
+type K8sAction string
+
+const (
+	K8sReadAction     K8sAction = "get"
+	K8sCreateAction   K8sAction = "create"
+	K8sUpdateAction   K8sAction = "patch"
+	K8sDeleteAction   K8sAction = "delete"
+	K8sWildcardAction K8sAction = "*"
+)
+
+var K8sActions = [5]K8sAction{
+	K8sReadAction,
+	K8sWildcardAction,
+	K8sCreateAction,
+	K8sUpdateAction,
+	K8sDeleteAction,
+}
+
+type K8sScope string
+
+const (
+	K8sWildcardScope   K8sScope = "*"
+	K8sDashboardScope  K8sScope = "persesdashboards"
+	K8sDatasourceScope K8sScope = "persesdatasources"
+	K8sProjectScope    K8sScope = "namespaces"
+)
+
+var K8sScopesToCheck = [2]K8sScope{
+	K8sDashboardScope,
+	K8sDatasourceScope,
+}
+
+type k8sUser struct {
+	UserInfo user.Info
+}
+
+// IsEnabled implements [Authorization]
+func (k *k8sImpl) IsEnabled() bool {
+	return true
+}
+
+// GetUser implements [Authorization]
+func (k *k8sImpl) GetUser(ctx echo.Context) (any, error) {
+	// Context can be nil when the function is called outside the request context.
+	// For example, the provisioning service is calling every service without any context.
+	if ctx == nil {
+		return nil, nil
+	}
+
+	if utils.IsAnonymous(ctx) {
+		return nil, nil
+	}
+
+	// At this point, we are sure that the context is not nil and the user is not anonymous.
+	// The user token is expected to be set in the Authorization header
+	req := ctx.Request().Clone(ctx.Request().Context())
+
+	// req.Header.Set("Authorization", GetAuthnHeaderFromClient(ctx, k.Kubeconfig))
+	res, ok, err := k.Authenticator.AuthenticateRequest(req)
+	if err != nil {
+		return nil, err
+	} else if !ok {
+		return nil, fmt.Errorf("request unable to be authenticated")
+	}
+
+	return k8sUser{UserInfo: res.User}, nil
+}
+
+// GetUsername implements [Authorization]
+func (k *k8sImpl) GetUsername(ctx echo.Context) (string, error) {
+	user, err := k.GetUser(ctx)
+	if err != nil {
+		return "", err
+	}
+	if user == nil {
+		return "", nil // No user found in the context, this is an anonymous endpoint
+	}
+	if k8sUser, ok := user.(k8sUser); ok {
+		return k8sUser.UserInfo.GetName(), nil
+	}
+	return "", apiInterface.InternalError
```

Comment:
can you put a comment, this case should not happen + an error log saying the type is not correct and printing the current type please ?